packages/toolpad-app/src/toolpadDataSources/googleSheets/client.tsx

```diff
@@ -1,149 +1,176 @@
 }
 
 function isConnectionValid(connection: GoogleSheetsConnectionParams | null): boolean {
-  if (connection?.access_token && connection?.expiry_date) {
-    if (connection.expiry_date >= Date.now()) {
-      return true;
-    }
+  if (connection?.access_token && connection?.refresh_token) {
+    return true;
   }
   return false;
 }
 
-const queryReducer = (
-  state: GoogleSheetsQuery,
-  action: GoogleSheetsQueryAction,
-): GoogleSheetsQuery => {
-  const { type } = action;
-  switch (type) {
-    case GoogleSheetsActionType.FETCH_SHEET:
-      return {
-        ...state,
-        sheet: null,
-        type,
-        spreadsheet: action.spreadsheet,
-      };
-    case GoogleSheetsActionType.UPDATE_SHEET:
-      return {
-        ...state,
-        type,
-        sheet: action.sheet,
-      };
-    case GoogleSheetsActionType.UPDATE_RANGE:
-      return {
-        ...state,
-        type,
-        ranges: action.ranges,
-      };
-    default:
-      return state;
-  }
-};
+function QueryEditor({ api, value, onChange }: QueryEditorProps<GoogleSheetsApiQuery>) {
+  const [spreadsheetQuery, setSpreadsheetQuery] = React.useState<string | null>(null);
+
+  const debouncedSpreadsheetQuery = useDebounced(spreadsheetQuery, 300);
 
-function QueryEditor({ value, onChange, api }: QueryEditorProps<GoogleSheetsQuery>) {
-  const [query, dispatch] = React.useReducer(
-    queryReducer,
-    value ?? {
-      type: GoogleSheetsActionType.FETCH_SPREADSHEETS,
-      spreadsheet: null,
-      sheet: null,
-      ranges: 'A1:Z100',
+  const fetchedFiles: PrivateApiResult<GoogleDriveFiles> = useQuery(
+    ['filesList', value?.spreadsheetId, debouncedSpreadsheetQuery],
+    () => {
+      return api.fetchPrivate({
+        type: GoogleSheetsPrivateQueryType.FILES_LIST,
+        spreadsheetQuery,
+      });
+    },
+    {
+      keepPreviousData: true,
     },
   );
 
-  const {
-    isIdle: isListPending,
-    isLoading: isListLoading,
-    data: listData,
-  } = useQuery('fetchSpreadsheets', () => {
-    return api.fetchPrivate({ type: GoogleSheetsActionType.FETCH_SPREADSHEETS });
-  });
-
-  const {
-    isIdle: isSpreadsheetPending,
-    isLoading: isSpreadsheetLoading,
-    data: spreadsheetData,
-  } = useQuery(
-    ['fetchSheet', query.spreadsheet?.id],
+  const fetchedFile: PrivateApiResult<GoogleDriveFile> = useQuery(
+    ['fileGet', value?.spreadsheetId],
     () => {
-      return api.fetchPrivate(query);
+      return api.fetchPrivate({
+        type: GoogleSheetsPrivateQueryType.FILE_GET,
+        spreadsheetId: value?.spreadsheetId,
+      });
     },
     {
-      enabled: Boolean(query.spreadsheet) && query.type === GoogleSheetsActionType.FETCH_SHEET,
+      keepPreviousData: true,
+      enabled: Boolean(value?.spreadsheetId),
     },
   );
 
-  const { isLoading: isApiDataLoading } = useQuery(
-    ['execApi', query.ranges],
+  const fetchedSpreadsheet: PrivateApiResult<GoogleSpreadsheet> = useQuery(
+    ['sheetGet', value?.spreadsheetId],
```

Comment:
Would it make sense to do

```suggestion
    ['sheets', GoogleSheetsPrivateQueryType.FETCH_SPREADSHEET, value?.spreadsheetId],
```

for these query keys?

Reply:
Yes, would it make sense then that the first element in the array is `sheets` for all of them?

Reply:
Not strictly necessary, but as Toolpad grows, we should take care that queries can't clash across the application. e.g. if tomorrow we support office 360 excel files or something there may be another `ExcelPrivateQueryType.FETCH_SPREADSHEET` constant with the same name. Prefixing the query key with something like the datasource name prevents these collisions from ever happening